client/src/connection_cache.rs

```diff
@@ -215,11 +217,49 @@ impl ConnectionCacheStats {
     }
 }
 
+pub const DEFAULT_CONNECTION_POOL_SIZE: usize = 4;
+
+/// Models the pool of connections
+struct ConnectionPool {
+    /// The connections in the pool
+    connections: Vec<Arc<Connection>>,
+
+    /// A simple reference count of connections being used.
+    /// A connection can be used multiple times. The reference count will be dropped
+    /// whe the connection object is dropped.
+    /// This helps to avoid creating new connections in case of the count of threads using
+    /// the pool is less than the pool size.
+    references: AtomicU64,
+
+    /// Connections in this pool share the same endpoint
+    endpoint: Arc<QuicLazyInitializedEndpoint>,
+}
+
+impl ConnectionPool {
+    /// Get a connection from the pool. It must have at least one connection in the pool.
+    /// This randomly picks a connection in the pool and increment the reference count.
+    fn borrow_connection(&self) -> Arc<Connection> {
+        let mut rng = thread_rng();
+        let n = rng.gen_range(0, self.connections.len());
+        let connection = self.connections[n].clone();
+        self.references.fetch_add(1, Ordering::Relaxed);
+        connection
+    }
+
+    /// Check if we need to create a new connection. If the count of the connections
+    /// is smaller than the pool size and we have outstanding users of the pool.
+    fn need_new_connection(&self, required_pool_size: usize) -> bool {
+        self.connections.len() < required_pool_size && self.references.load(Ordering::Relaxed) > 0
```

Comment:
I think we might need to use acquire ordering here (https://stackoverflow.com/questions/27751025/why-is-an-acquire-barrier-needed-before-deleting-the-data-in-an-atomically-refer). In any case, I'm a little uncomfortable with seemingly reinventing/reimplementing the wheel with reference-counted resources, which is notoriously hard to get right and fast for us mortals 😅...

Reply:
Yes. It is tricky -- that reference count is an optimization. And it turns out the optimization is premature and ineffective in our use case as it is very easy to have multiple threads using the connections in our case driven from the runtime. I have removed all reference counting logic!